File: DP/projects.py
from bisect import bisect_left

n = int(input())

arr = []

for _ in range(n):
    start, end, reward = map(int, input().split())
    arr.append((end, start, reward))

arr.sort()

mp = {0:0}
maxReward = 0

for each in arr:
    end, start, reward  = each
    a = list(mp.keys())

    pos = bisect_left(a, start)
    pos -= 1

    totalReward = mp[a[pos]] + reward
    maxReward = max(totalReward, maxReward)
    mp[end] = maxReward

print(maxReward)

File: DP/projects.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

#define ll long long

int main() {
    // read the input
    int n;
    std::cin >> n;
    std::vector<std::tuple<ll, ll, ll>> projects(n);

    for (int i = 0; i < n; ++i) {
        int start, end, reward;
        std::cin >> start >> end >> reward;
        projects[i]={end, start, reward};
    }

    std::sort(projects.begin(), projects.end());

    // compute the dp table and totalBestReward
    ll maxReward = 0;
    std::map<ll, ll> dp;  // {end: totalRewardSoFar}
    dp[0] = 0;

    for (const auto& project : projects) {
        ll end, start, reward;
        std::tie(end, start, reward) = project;

        auto it = dp.lower_bound(start);
        it--;
        ll totalReward = it->second + reward;
        maxReward = std::max(maxReward, totalReward);
        dp[end] = maxReward;
    }

    std::cout << maxReward << std::endl;

    return 0;
}
